Add text covering the option type (#633)
ocaml/v3.ocaml.org: data/tutorials/lg_00_data_types_and_matching.md

`option` type. Like `list`, it's predefined. Here is how:

<!-- $MDX non-deterministic=command -->
```ocaml
# #show option;;
type 'a option = None | Some of 'a
```

`option` used to wrap some data, if available, or absence of data otherwise.
Here is 42, stored inside an `option` using the data carrying constructor
`Some`:

```ocaml
# Some 42;;
- : int option = Some 42
```

The `None` constructor means no data is availble.

In other words a value of type `t option` for some type `t` represents:

* either a value `v` of type `t`, wrapped as `Some v`,
* no such value, then `o` has the value `None`.

The option type is very useful when lack of data is better handled as a special
value (_i.e._ `None`) rather than an exception. It is the type-safe version of
returning error values such as in C, for instance. Since no data has any special
meaning, confusion between regular values and absence of value is impossible. In
computer science, this type is called the [option
type](https://en.wikipedia.org/wiki/Option_type). OCaml has supported `option`
since its inception.

The function `Sys.getenv : string -> string` from the standard library allows to

> query the value of an environment variable, however the it throws an exception

Reviewer: ```suggestion
query the value of an environment variable, however, it throws an exception
```

if the variable is not defined. On the other hand, the function `Sys.getenv_opt
: string -> string opt` does the same except it returns `None` is the variable

> is not defined. Here is what may happen on a machine where a variable called
> `EDITOR` is defined:

Reviewer: ```suggestion
is not defined. Here is what may happen if we try to access an undefined environment variable:
```

```ocaml

> # Sys.getenv "DUMMY";;
> Exception: Not_found.
> # Sys.getenv_opt "DUMMY";;
> - : string option = None

Reviewer: ```suggestion
# Sys.getenv "UNDEFINED_ENVIRONMENT_VARIABLE";;
Exception: Not_found.
# Sys.getenv_opt "UNDEFINED_ENVIRONMENT_VARIABLE";;
- : string option = None
```

> Similar functions can be written to look up values in a dictionary, to convert
> a list of pairs to or from a tree dictionary and so on.

Reviewer 2: ```suggestion
Similar functions can be written to look up values in a dictionary, to convert
a list of pairs to or from a tree dictionary, and so on.
```

> ## Example: options

Reviewer 2: ```suggestion
## Example: Options
```

The OCaml language has been supporting the [option

> type](https://en.wikipedia.org/wiki/Option_type) since its inception.

Reviewer: The Wikipedia article reads:

> In [programming languages](https://en.wikipedia.org/wiki/Programming_language) (especially [functional programming](https://en.wikipedia.org/wiki/Functional_programming) languages) and [type theory](https://en.wikipedia.org/wiki/Type_theory), an option type or maybe type is a [polymorphic type](https://en.wikipedia.org/wiki/Parametric_polymorphism) that represents encapsulation of an optional value; e.g., it is used as the return type of functions which may or may not return a meaningful value when they are applied. It consists of a constructor which either is empty (often named None or Nothing), or which encapsulates the original data type A (often written Just A or Some A).

I was expecting something much more complicated, but still, I'm not sure it's the most beginner-friendly explanation we could give. E.g. this assumes that the reader knows what encapsulation means (and coming from OOP, what one understands by "encapsulation" can be something completely different!). Same for polymorphic types. I'd recommend writing a friendlier introduction of the option type in the doc.

<!-- $MDX non-deterministic=command -->
```ocaml
# #show option;;
type 'a option = None | Some of 'a
```

A value of type `t option` for some type `t` represent either a value `v` of

> type `t`, wraped as `Some v`, or no such value, in such case `o` has the value

Reviewer 2: ```suggestion
type `t`, wrapped as `Some v`, or no such value, in such case `o` has the value
```

`None`. The option type is very useful when lack of data is better handled as a

> special value (_i.e._ `None`) rather than an exception. It is the type-safe

Reviewer 2: ```suggestion
special value (_i.e.,_ `None`) rather than an exception. It is the type-safe
```

version of returning error values such as in C, for instance.

> The function `Sys.getenv_opt : string -> string opt` from the standard library
> allows to query the value of an environment variable from the running process.
> Here is what may happen on a machine where a variable called `EDITOR` is
> defined:

Reviewer 2: ```suggestion
The function `Sys.getenv_opt : string -> string opt` from the standard library
allows users to query the value of an environment variable from the running process.
Here is what may happen on a machine where a variable called `EDITOR` is
defined:
```

<!-- $MDX non-deterministic=output -->
```ocaml
# Sys.getenv_opt "EDITOR";;
- : string option = Some "vi"
```

> However if no such variable is defined the returned value will be `None`.

Reviewer 2: ```suggestion
However if no such variable is defined, the returned value will be `None`.
```

> Using pattern-matching, it is possible to define functions allowing to easily

Reviewer 2: ```suggestion
Using pattern-matching, it is possible to define functions, allowing users to easily
```

> work with option values. Here is `map` which allows to apply a function to the

Reviewer 2: ```suggestion
work with option values. Here is `map` which allows the application of a function to the
```

value wrapped inside an option, if present:

```ocaml
# let map f = function
  | None -> None
  | Some v -> Some (f v);;
val map : ('a -> 'b) -> 'a option -> 'b option = <fun>
```

> Here is `join`, which peels off one layer from a doubly wrapped option:

Reviewer: I found that using the types when explaining functions can aid the reader/learner's understanding. For instance, the above can seem a bit abstract, but reading `'a option option -> 'a option` at the same time might help.

```ocaml
# let join = function
  | Some o -> o
  | None -> None;;
val join : 'a option option -> 'a option = <fun>
```

> Here are generic functions allowing to consume and produce option values. `fold`

Reviewer 2: ```suggestion
Here are generic functions allowing users to consume and produce option values. `fold`
```

takes an option parameter, whilst `unfold` returns an option value.

```ocaml
# let fold f default = function
  | Some v -> f v
  | None -> default
  let unfold p f x =
    if p x then
      Some (f x)
    else
      None;;
val fold : ('a -> 'b) -> 'b -> 'a option -> 'b = <fun>
val unfold : ('a -> bool) -> ('a -> 'b) -> 'a -> 'b option = <fun>
```

Most of those functions as well as other useful ones are provided by the

> Standard Library in the [`Stdlib.Option`](https://ocaml.org/api/Option.html)

Reviewer 2: You have "Standard Library" capitalised here but not earlier in the doc. Which is correct?

Author: Nope :-) It should be all the same way. I'll fix that.

supporting module.

> 

Reviewer: Could be a nice addition after these, to add something like "oh, by the way, a module that defines the functions above is called a monad, but we'll get back to it".
